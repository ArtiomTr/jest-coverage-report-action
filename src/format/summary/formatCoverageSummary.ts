import table from 'markdown-table';

import { CoverageSummary } from '../../typings/Coverage';
import { formatPercentage } from '../../utils/formatPercentage';
import { getStatusOfPercents } from '../../utils/getStatusOfPercents';
import { i18n } from '../../utils/i18n';
import { withExplanation } from '../../utils/withExplanation';

export const formatCoverageSummary = (
    headSummary: Array<CoverageSummary>,
    baseSummary: Array<CoverageSummary> | undefined,
    threshold: number | undefined
): string =>
    table(
        [
            [
                withExplanation(i18n('status'), i18n('statusExplanation')),
                i18n('category'),
                i18n('percentage'),
                i18n('ratio'),
            ],
            ...headSummary.map((currSummary, index) => [
                getStatusOfPercents(currSummary.percentage, threshold),
                currSummary.title,
                formatPercentage(
                    currSummary.percentage,
                    baseSummary?.[index].percentage
                ),
                `${currSummary.covered}/${currSummary.total}`,
            ]),
        ],
        { align: ['c', 'l', 'l', 'c'] }
    );
