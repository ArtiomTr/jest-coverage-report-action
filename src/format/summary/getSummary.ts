import { CoverageSummary } from '../../typings/Coverage';
import { CoverageMap, FileCoverage } from '../../typings/JsonReport';
import { getPercents } from '../getPercents';

export const getSummary = (
    map: CoverageMap,
    totalCounter: (value: FileCoverage) => number,
    coveredCounter: (value: FileCoverage) => number,
    title: string
): CoverageSummary => {
    const total = Object.values(map).reduce(
        (acc, currValue) => acc + totalCounter(currValue),
        0
    );

    const covered = Object.values(map).reduce(
        (acc, currValue) => acc + coveredCounter(currValue),
        0
    );

    return {
        title,
        total,
        covered,
        percentage: getPercents(covered, total),
    };
};
