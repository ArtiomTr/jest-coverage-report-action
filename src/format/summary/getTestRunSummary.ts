import { JsonReport } from '../../typings/JsonReport';
import { i18n } from '../../utils/i18n';

export const getTestRunSummary = (jsonReport: JsonReport) =>
    jsonReport.success
        ? i18n('testsSuccessSummary', {
              numPassedTests: jsonReport.numPassedTests,
              numPassedTestSuites: jsonReport.numPassedTestSuites,
              ending: jsonReport.numPassedTestSuites > 1 ? 's' : '',
          })
        : i18n('testsFailSummary', {
              numFailedTests: jsonReport.numFailedTests,
              numTotalTests: jsonReport.numTotalTests,
              numFailedTestSuites: jsonReport.numFailedTestSuites,
              numTotalTestSuites: jsonReport.numTotalTestSuites,
          });
