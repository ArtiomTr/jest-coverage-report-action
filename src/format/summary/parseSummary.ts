import { getSummary } from './getSummary';
import { JsonReport } from '../../typings/JsonReport';
import { i18n } from '../../utils/i18n';
import {
    coveredBranchesCounter,
    coveredLinesCounter,
    standardCoveredCounter,
    standardTotalCounter,
    totalBranchesCounter,
    totalLinesCounter,
} from '../counters';

export const parseSummary = (jsonReport: JsonReport) => {
    return [
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('s'),
            standardCoveredCounter('s'),
            i18n('statements')
        ),
        getSummary(
            jsonReport.coverageMap,
            totalBranchesCounter,
            coveredBranchesCounter,
            i18n('branches')
        ),
        getSummary(
            jsonReport.coverageMap,
            standardTotalCounter('f'),
            standardCoveredCounter('f'),
            i18n('functions')
        ),
        getSummary(
            jsonReport.coverageMap,
            totalLinesCounter,
            coveredLinesCounter,
            i18n('lines')
        ),
    ];
};
