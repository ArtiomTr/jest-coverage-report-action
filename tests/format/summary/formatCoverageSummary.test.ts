import { formatCoverageSummary } from '../../../src/format/summary/formatCoverageSummary';

describe('formatCoverageSummary', () => {
    it('should format summary when head and base coverage specified', () => {
        expect(
            formatCoverageSummary(
                [
                    {
                        title: 'Statements',
                        total: 30,
                        covered: 23,
                        percentage: 76.67,
                    },
                ],
                [
                    {
                        title: 'Statements',
                        total: 30,
                        covered: 15,
                        percentage: 50,
                    },
                ],
                70
            )
        ).toMatchSnapshot();

        expect(
            formatCoverageSummary(
                [
                    {
                        title: 'Statements',
                        total: 30,
                        covered: 23,
                        percentage: 76.67,
                    },
                ],
                [
                    {
                        title: 'Statements',
                        total: 30,
                        covered: 29,
                        percentage: 96.67,
                    },
                ],
                70
            )
        ).toMatchSnapshot();
    });

    it('should format summary when head and base coverage specified (with threshold)', () => {
        expect(
            formatCoverageSummary(
                [
                    {
                        title: 'Statements',
                        total: 30,
                        covered: 23,
                        percentage: 76.67,
                    },
                ],
                [
                    {
                        title: 'Statements',
                        total: 30,
                        covered: 15,
                        percentage: 50,
                    },
                ],
                70
            )
        ).toMatchSnapshot();

        expect(
            formatCoverageSummary(
                [
                    {
                        title: 'Statements',
                        total: 30,
                        covered: 23,
                        percentage: 76.67,
                    },
                ],
                [
                    {
                        title: 'Statements',
                        total: 30,
                        covered: 29,
                        percentage: 96.67,
                    },
                ],
                80
            )
        ).toMatchSnapshot();
    });

    it('should format summary when head and base coverage specified (with threshold)', () => {
        expect(
            formatCoverageSummary(
                [
                    {
                        title: 'Statements',
                        total: 30,
                        covered: 23,
                        percentage: 76.67,
                    },
                ],
                undefined,
                90
            )
        ).toMatchSnapshot();

        expect(
            formatCoverageSummary(
                [
                    {
                        title: 'Statements',
                        total: 30,
                        covered: 29,
                        percentage: 96.67,
                    },
                ],
                undefined,
                90
            )
        ).toMatchSnapshot();
    });
});
