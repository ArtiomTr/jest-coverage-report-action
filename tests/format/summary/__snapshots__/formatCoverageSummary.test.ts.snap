// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatCoverageSummary > should format summary when head and base coverage specified (with threshold) 1`] = `
"| <div title=\\"Status of coverage:&#10; 游릭 - ok&#10; 游리 - slightly more than threshold&#10; 游댮 - under the threshold\\">St.<sup>:grey_question:</sup></div> | Category   | Percentage                                                   | Covered / Total |
| :----------------------------------------------------------------------------------------------------------------------------------------------------: | :--------- | :----------------------------------------------------------- | :-------------: |
|                                                                           游리                                                                           | Statements | <div title=\\"Base coverage is: 50%\\">76.67% (+26.67% 游댶)</div> |      23/30      |"
`;

exports[`formatCoverageSummary > should format summary when head and base coverage specified (with threshold) 2`] = `
"| <div title=\\"Status of coverage:&#10; 游릭 - ok&#10; 游리 - slightly more than threshold&#10; 游댮 - under the threshold\\">St.<sup>:grey_question:</sup></div> | Category   | Percentage                                                   | Covered / Total |
| :----------------------------------------------------------------------------------------------------------------------------------------------------: | :--------- | :----------------------------------------------------------- | :-------------: |
|                                                                           游댮                                                                           | Statements | <div title=\\"Base coverage is: 96.67%\\">76.67% (-20% 游댵)</div> |      23/30      |"
`;

exports[`formatCoverageSummary > should format summary when head and base coverage specified (with threshold) 3`] = `
"| <div title=\\"Status of coverage:&#10; 游릭 - ok&#10; 游리 - slightly more than threshold&#10; 游댮 - under the threshold\\">St.<sup>:grey_question:</sup></div> | Category   | Percentage | Covered / Total |
| :----------------------------------------------------------------------------------------------------------------------------------------------------: | :--------- | :--------- | :-------------: |
|                                                                           游댮                                                                           | Statements | 76.67%     |      23/30      |"
`;

exports[`formatCoverageSummary > should format summary when head and base coverage specified (with threshold) 4`] = `
"| <div title=\\"Status of coverage:&#10; 游릭 - ok&#10; 游리 - slightly more than threshold&#10; 游댮 - under the threshold\\">St.<sup>:grey_question:</sup></div> | Category   | Percentage | Covered / Total |
| :----------------------------------------------------------------------------------------------------------------------------------------------------: | :--------- | :--------- | :-------------: |
|                                                                           游릭                                                                           | Statements | 96.67%     |      29/30      |"
`;

exports[`formatCoverageSummary > should format summary when head and base coverage specified 1`] = `
"| <div title=\\"Status of coverage:&#10; 游릭 - ok&#10; 游리 - slightly more than threshold&#10; 游댮 - under the threshold\\">St.<sup>:grey_question:</sup></div> | Category   | Percentage                                                   | Covered / Total |
| :----------------------------------------------------------------------------------------------------------------------------------------------------: | :--------- | :----------------------------------------------------------- | :-------------: |
|                                                                           游리                                                                           | Statements | <div title=\\"Base coverage is: 50%\\">76.67% (+26.67% 游댶)</div> |      23/30      |"
`;

exports[`formatCoverageSummary > should format summary when head and base coverage specified 2`] = `
"| <div title=\\"Status of coverage:&#10; 游릭 - ok&#10; 游리 - slightly more than threshold&#10; 游댮 - under the threshold\\">St.<sup>:grey_question:</sup></div> | Category   | Percentage                                                   | Covered / Total |
| :----------------------------------------------------------------------------------------------------------------------------------------------------: | :--------- | :----------------------------------------------------------- | :-------------: |
|                                                                           游리                                                                           | Statements | <div title=\\"Base coverage is: 96.67%\\">76.67% (-20% 游댵)</div> |      23/30      |"
`;
